### ISSUE TITLE
Different behaviour depending on service account settings even when kr8s should ignore these

### ISSUE THREAD

**Issue by gazzar** [kr8s-org/kr8s#234]
### Which project are you reporting a bug for?

kr8s

### What happened?

kr8s is giving me errors in one of my container environments, where it behaves differently to kubectl run directly in the same container. Because kubectl behaves differently, I'm reporting this as a bug, although I'm not actually certain that it is a bug in kr8s; it could be a configuration error in my environment or just a misunderstanding about authentication. In a different environment, where the code was developed, my code that uses kr8s runs fine, the difference being that no service account is configured there.

The start of the traceback in the problematic environment is as follows - the code in pods.py below is within a FastAPI endpoint, whereas the kr8s api is configured in the FastAPI startup method, described below

```
2023-12-07 00:33:50,340 - ERROR - pods.py - fastapi - Failed to list the pods with exception pods is forbidden: User "system:serviceaccount:space-mct:mct-service-status-api-dev" cannot list resource "pods" in API group "" in the namespace "space-mct"
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/kr8s/_api.py", line 136, in call_api
    response.raise_for_status()
  File "/usr/local/lib/python3.11/site-packages/httpx/_models.py", line 749, in raise_for_status
    raise HTTPStatusError(message, request=request, response=self)
httpx.HTTPStatusError: Client error '403 Forbidden' for url 'https://10.26.0.1/api/v1/namespaces/space-mct/pods'
For more information check: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/src/app/api/routes/pods.py", line 130, in get_pods_in_namespace
    for pod in await kr8s.asyncio.get("pods", namespace=namespace):
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/kr8s/asyncio/_helpers.py", line 65, in get
    return await api._get(
           ^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/kr8s/_api.py", line 341, in _get
    async with self._get_kind(
  File "/usr/local/lib/python3.11/contextlib.py", line 210, in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/kr8s/_api.py", line 270, in _get_kind
    async with self.call_api(
  File "/usr/local/lib/python3.11/contextlib.py", line 210, in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/kr8s/_api.py", line 149, in call_api
    raise ServerError(
kr8s._exceptions.ServerError: pods is forbidden: User "system:serviceaccount:space-mct:mct-service-status-api-dev" cannot list resource "pods" in API group "" in the namespace "space-mct"
2023-12-07 00:33:50,340 - ERROR - pods.py - fastapi - Failed to list the pods with exception pods is forbidden: User "system:serviceaccount:space-mct:mct-service-status-api-dev" cannot list resource "pods" in API group "" in the namespace "space-mct"
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/kr8s/_api.py", line 136, in call_api
    response.raise_for_status()
  File "/usr/local/lib/python3.11/site-packages/httpx/_models.py", line 749, in raise_for_status
    raise HTTPStatusError(message, request=request, response=self)
httpx.HTTPStatusError: Client error '403 Forbidden' for url 'https://10.26.0.1/api/v1/namespaces/space-mct/pods'
```

The kr8s auth.token and auth.server settings are defined in a .kubeconfig file, which are the ones I want to use. In the environment that exhibits the problem, there is a service account configured via KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT environment variables, but I'm trying to avoid using that. Code using the "old" vanilla kubernetes api and kubectl both behave fine in this environment, but my kr8s functions break, possibly because kr8s attempts to authenticate against the service account. I'm not sure exactly what's going on to trigger the failure.
As an attempted workaround for this, I tried directly calling the _load_kubeconfig() method in my startup code as follows:

```
client = kr8s.api(kubeconfig=KUBECONFIG, serviceaccount=None)
print(client.auth.server)
await client.auth._load_kubeconfig()
print(client.auth.server)
```

I added this because the api() method call doesn't run _load_kubeconfig() in the environment with the service account. Looking at the kr8s _auth.py reathenticate() method logic, this is because _load_service_account() populates self.server and thus skips the subsequent _load_kubeconfig(). The kr8s docs say the lookup order is:
  - ~/.kube/config (or the path set by the KUBECONFIG environment variable)
  - /var/run/secrets/kubernetes.io/serviceaccount
but this seems not to match what actually happens, and having a service account configured seems to preclude specifying anything else in the .kubeconfig, which is not what I expected. Calling _load_kubeconfig() directly updates auth.token and auth.server from the .kubeconfig, but this doesn't actually fix things; it seems that kr8s continues to attempt to authenticate against the service account. The two print statements in the startup code snippet output

```
[https://10.26.0.1:443](https://10.26.0.1/)
https://api.imgnetes.beamline.synchrotron.org.au:6443/
```

and I would expect kr8s to use the updated server details on subsequent calls until something causes a reauthentication, but as you can see from the traceback, it reverts to using the service host address.
I couldn't find anything in the kubernetes docs that define the expected lookup order, so I'm not sure if the kr8s authentication is implemented correctly or not, or what else I can look at to understand the problem.

**Comment by jacobtomlinson** [comment]
Thanks for raising this @gazzar. I agree that the code doesn't match the documentation.

I've opened #237 with a potential fix. Would you mind installing it and checking that it resolves your problem?

```
pip install git+https://github.com/jacobtomlinson/kr8s.git@fix-auth-order
```

**Comment by gazzar** [comment]
Thank you @jacobtomlinson for looking into this so quickly. I'll have to wait until next week (hopefully first thing) to try this and let you know whether it works in both my environments.

**Comment by gazzar** [comment]
Hi @jacobtomlinson 
Good news. Preliminary testing shows that the problem I described is resolved by this change in my dev environment. Tomorrow I want to also check it in the environment that was already working to verify that still behaves as before.

**Comment by gazzar** [comment]
@jacobtomlinson I can now confirm that your authentication-order fix is working well in all my environments and has not caused any regressions. Thanks again for following up on this so quickly.

**Comment by jacobtomlinson** [comment]
Great thanks for confirming @gazzar. I've just released `v0.12.6` which contains this fix.

**Comment by gazzar** [comment]
Amazing speed again! Thanks @jacobtomlinson !
